### ISSUE TITLE
[QUERY] Token refresh/Connection issues with Service Bus Client

### ISSUE THREAD

**Issue by neumannm** [Azure/azure-sdk-for-java#39152]
**Query/Question**
I am having several issues related to some token refresh and connection problems that I don't know how to fix. I am hoping to get some help here.

**Scenario:** We have developed a small application that helps us manage several queues in our Service Bus namespace. It can list some stats about the queues (no. of messages per queue, also in DLQ), it allows us to inspect the messages and also to delete or restart them.
All of this functionality is triggered via REST requests, e.g. from our UI.

**Problem:** We observe some strange behaviour with our application. From time to time, our requests from the UI receive a HTTP 504 (Gateway Timeout), where the request to list messages in a queue yields no result. Then we have to wait some time to try again. Also, we can see some error messages in the log.

**Example:** So I trigger the endpoint once. This statement in the logs appears everytime one request is processed:
```
2024-03-11 11:09:07.279  WARN 1 --- [tor-executor-20] c.a.m.s.ServiceBusMessageSerializer      : {"az.sdk.message":"AMQP response did not contain OK status code.","statusCode":"NO_CONTENT"} 
```

Now, *exactly 18 Minutes later* (reproducible), we get the following 2 error messages in the log:
```
2024-03-11 11:27:06.017 ERROR 1 --- [     parallel-2] c.a.c.a.i.AmqpChannelProcessor           : {"az.sdk.message":"Cannot subscribe. Processor is already terminated.","exception":"Cannot subscribe. Processor is already terminated.","connectionId":"MF_094acf_1710151746979","entityPath":"$cbs"}
2024-03-11 11:27:06.017 ERROR 1 --- [     parallel-2] c.a.c.a.i.ActiveClientTokenManager       : {"az.sdk.message":"Error occurred while refreshing token that is not retriable. Not scheduling refresh task. Use ActiveClientTokenManager.authorize() to schedule task again.","exception":"Cannot subscribe. Processor is already terminated.","scopes":"amqp://my-sbns.servicebus.windows.net/MyQueue/$deadletterqueue","audience":"amqp://my-sbns.servicebus.windows.net/MyQueue/$deadletterqueue"}
```

Still, making more requests is possible without problems, until it isn't. So for these 5 requests I got a result:
```
2024-03-11 11:27:44.331  WARN 1 --- [tor-executor-21] c.a.m.s.ServiceBusMessageSerializer      : {"az.sdk.message":"AMQP response did not contain OK status code.","statusCode":"NO_CONTENT"}
2024-03-11 11:29:00.927  WARN 1 --- [tor-executor-22] c.a.m.s.ServiceBusMessageSerializer      : {"az.sdk.message":"AMQP response did not contain OK status code.","statusCode":"NO_CONTENT"}
2024-03-11 11:29:03.882  WARN 1 --- [tor-executor-23] c.a.m.s.ServiceBusMessageSerializer      : {"az.sdk.message":"AMQP response did not contain OK status code.","statusCode":"NO_CONTENT"}
2024-03-11 11:29:07.031  WARN 1 --- [tor-executor-24] c.a.m.s.ServiceBusMessageSerializer      : {"az.sdk.message":"AMQP response did not contain OK status code.","statusCode":"NO_CONTENT"}
2024-03-11 11:29:09.481  WARN 1 --- [tor-executor-25] c.a.m.s.ServiceBusMessageSerializer      : {"az.sdk.message":"AMQP response did not contain OK status code.","statusCode":"NO_CONTENT"}
```

More or less exactly *1 Minute* after the (6th) request that did not yield a result, we get the HTTP 504 and see the following in the logs:
```
2024-03-11 11:33:30.449  WARN 1 --- [tor-executor-30] c.a.m.s.ServiceBusMessageSerializer      : {"az.sdk.message":"AMQP response did not contain OK status code.","statusCode":"NO_CONTENT"}

2024-03-11 11:34:32.309  WARN 1 --- [tor-executor-31] c.a.c.a.i.handler.ConnectionHandler      : {"az.sdk.message":"onTransportError","connectionId":"MF_aa63b1_1710153212307","errorCondition":"amqp:connection:framing-error","errorDescription":"org.apache.qpid.proton.engine.TransportException: connection aborted","hostName":"my-sbns.servicebus.windows.net"}
2024-03-11 11:34:32.310 ERROR 1 --- [     parallel-2] reactor.core.publisher.Operators         : Operator called default onErrorDropped

reactor.core.Exceptions$ErrorCallbackNotImplemented: com.azure.messaging.servicebus.ServiceBusException: Connection 'MF_aa63b1_1710153212307' not opened within AmqpRetryOptions.tryTimeout(): PT1M, errorContext[NAMESPACE: my-sbns.servicebus.windows.net. ERROR CONTEXT: N/A]
Caused by: com.azure.messaging.servicebus.ServiceBusException: Connection 'MF_aa63b1_1710153212307' not opened within AmqpRetryOptions.tryTimeout(): PT1M, errorContext[NAMESPACE: my-sbns.servicebus.windows.net. ERROR CONTEXT: N/A]
        at com.azure.messaging.servicebus.ServiceBusReceiverAsyncClient.mapError(ServiceBusReceiverAsyncClient.java:1670) ~[azure-messaging-servicebus-7.14.4.jar:7.14.4]
        at com.azure.messaging.servicebus.ServiceBusReceiverAsyncClient.lambda$peekMessages$14(ServiceBusReceiverAsyncClient.java:738) ~[azure-messaging-servicebus-7.14.4.jar:7.14.4]
        at reactor.core.publisher.Flux.lambda$onErrorMap$28(Flux.java:7070) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:94) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.MonoFlatMapMany$FlatMapManyMain.onError(MonoFlatMapMany.java:204) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.MonoFlatMap$FlatMapMain.secondError(MonoFlatMap.java:192) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.MonoFlatMap$FlatMapInner.onError(MonoFlatMap.java:259) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.onError(MonoIgnoreThen.java:278) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.MonoPeekTerminal$MonoTerminalPeekSubscriber.onError(MonoPeekTerminal.java:258) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.FluxHide$SuppressFuseableSubscriber.onError(FluxHide.java:142) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.onError(MonoIgnoreThen.java:278) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.SerializedSubscriber.onError(SerializedSubscriber.java:124) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.SerializedSubscriber.onError(SerializedSubscriber.java:124) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.FluxTimeout$TimeoutOtherSubscriber.onError(FluxTimeout.java:341) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.Operators.error(Operators.java:198) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.MonoErrorSupplied.subscribe(MonoErrorSupplied.java:56) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.Mono.subscribe(Mono.java:4490) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.handleTimeout(FluxTimeout.java:301) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.doTimeout(FluxTimeout.java:280) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.FluxTimeout$TimeoutTimeoutSubscriber.onNext(FluxTimeout.java:419) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.FluxOnErrorReturn$ReturnSubscriber.onNext(FluxOnErrorReturn.java:162) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.MonoDelay$MonoDelayRunnable.propagateDelay(MonoDelay.java:271) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.MonoDelay$MonoDelayRunnable.run(MonoDelay.java:286) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28) ~[reactor-core-3.4.28.jar:3.4.28]
        at java.base/java.util.concurrent.FutureTask.run(Unknown Source) ~[na:na]
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[na:na]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[na:na]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[na:na]
        at java.base/java.lang.Thread.run(Unknown Source) ~[na:na]
Caused by: com.azure.core.amqp.exception.AmqpException: Connection 'MF_aa63b1_1710153212307' not opened within AmqpRetryOptions.tryTimeout(): PT1M, errorContext[NAMESPACE: my-sbns.servicebus.windows.net. ERROR CONTEXT: N/A]
        at com.azure.core.amqp.implementation.ReactorConnection.lambda$new$1(ReactorConnection.java:137) ~[azure-core-amqp-2.8.9.jar:2.8.9]
        at reactor.core.publisher.MonoErrorSupplied.subscribe(MonoErrorSupplied.java:55) ~[reactor-core-3.4.28.jar:3.4.28]
        ... 14 common frames omitted

2024-03-11 11:34:32.313 ERROR 1 --- [nio-8080-exec-5] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed; nested exception is com.azure.messaging.servicebus.ServiceBusException: Connection 'MF_aa63b1_1710153212307' not opened within AmqpRetryOptions.tryTimeout(): PT1M, errorContext[NAMESPACE: my-sbns.servicebus.windows.net. ERROR CONTEXT: N/A]] with root cause

com.azure.core.amqp.exception.AmqpException: Connection 'MF_aa63b1_1710153212307' not opened within AmqpRetryOptions.tryTimeout(): PT1M, errorContext[NAMESPACE: my-sbns.servicebus.windows.net. ERROR CONTEXT: N/A]
        at com.azure.core.amqp.implementation.ReactorConnection.lambda$new$1(ReactorConnection.java:137) ~[azure-core-amqp-2.8.9.jar:2.8.9]
        at reactor.core.publisher.MonoErrorSupplied.subscribe(MonoErrorSupplied.java:55) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.Mono.subscribe(Mono.java:4490) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.handleTimeout(FluxTimeout.java:301) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.FluxTimeout$TimeoutMainSubscriber.doTimeout(FluxTimeout.java:280) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.FluxTimeout$TimeoutTimeoutSubscriber.onNext(FluxTimeout.java:419) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.FluxOnErrorReturn$ReturnSubscriber.onNext(FluxOnErrorReturn.java:162) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.MonoDelay$MonoDelayRunnable.propagateDelay(MonoDelay.java:271) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.publisher.MonoDelay$MonoDelayRunnable.run(MonoDelay.java:286) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68) ~[reactor-core-3.4.28.jar:3.4.28]
        at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28) ~[reactor-core-3.4.28.jar:3.4.28]
        at java.base/java.util.concurrent.FutureTask.run(Unknown Source) ~[na:na]
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[na:na]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[na:na]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[na:na]
        at java.base/java.lang.Thread.run(Unknown Source) ~[na:na]
```

The error from the `ActiveClientTokenManager` now starts to re-appear in the log several times, I suspect one for each of the previous requests.
```
2024-03-11 11:45:42.962 ERROR 1 --- [     parallel-4] c.a.c.a.i.AmqpChannelProcessor           : {"az.sdk.message":"Cannot subscribe. Processor is already terminated.","exception":"Cannot subscribe. Processor is already terminated.","connectionId":"MF_1fd926_1710152855780","entityPath":"$cbs"}
2024-03-11 11:45:42.962 ERROR 1 --- [     parallel-4] c.a.c.a.i.ActiveClientTokenManager       : {"az.sdk.message":"Error occurred while refreshing token that is not retriable. Not scheduling refresh task. Use ActiveClientTokenManager.authorize() to schedule task again.","exception":"Cannot subscribe. Processor is already terminated.","scopes":"amqp://my-sbns.servicebus.windows.net/MyQueue/$deadletterqueue","audience":"amqp://my-sbns.servicebus.windows.net/MyQueue/$deadletterqueue"}
```

Note that I am not sure that this error and the "Connection not opened" thing are even related.

***Why is this not a Bug or a feature Request?***
Might be a bug as well, but I cannot rule out that it's actually our implementation's (library usage) fault.

**Setup (please complete the following information if applicable):**
 - Library/Libraries: com.azure:azure-messaging-servicebus:7.14.4

Here's the code we use to peek at messages in a specific queue. I have commented some lines for clarification and for places where I am unsure about correct library usage:
```java
final LocalDateTime startTime = LocalDateTime.now();
final List<QueueMessage> queueMessages = new ArrayList<>();

// hint: clientBuilder is a Spring bean (new ServiceBusClientBuilder().connectionString(connectionString).enableCrossEntityTransactions())
try (ServiceBusReceiverClient receiver = clientBuilder
	.receiver()
    // we only peek and do not really process messages, so 1 minute max timeout should be enough (?)
	.maxAutoLockRenewDuration(Duration.ofMinutes(1L))
	.queueName(queueName)
    // we only peek and to not want to change message state - is this even needed when using peekMessages only?
	.disableAutoComplete()
	.buildClient()) {
        loop:
		while (true) {
            // peek next (up to) 100 messages
			final IterableStream<ServiceBusReceivedMessage> receivedMessages = receiver.peekMessages(100);
			if (receivedMessages.stream().findAny().isEmpty()) {
                // no (more) messages
				break;
			}
			for (ServiceBusReceivedMessage receivedMessage : receivedMessages) {
				final QueueMessage queueMessage = AzureQueueMapper.parseMessage(receivedMessage);
                /*
                * idea: to prevent infinite loop
                * 1) stop as soon as we see a message that was enqueued after we started peeking
                * 2) stop as soon as we have reached the configured max number of messages to peek
                */
				if (queueMessage.getCreationDate().isAfter(startTime) || queueMessages.size() == maxEntries) {
					break loop;
				}
				queueMessages.add(queueMessage);
			}
		}
		return queueMessages;
}
```

**Things I am not sure about in the implementation:** 
* `enableCrossEntityTransactions()` causing problems? We use this because in other methods we also re-send messages from a DLQ to the "normal" queue for reprocessing, which means we both need a receiver for the DLQ and a sender for the "normal" queue.
* `.maxAutoLockRenewDuration(Duration.ofMinutes(1L))`? The Message lock duration is also set to 1 Minute on the queue.
* `.disableAutoComplete()` needed when we only peek?
* is there a better way to achieve what we want, i.e. listing all messages currently present in a queue without processing them?

**Comment by joshfree** [comment]
@anuchandy please take a look

**Comment by anuchandy** [comment]
Hello @neumannm,


1.	There shouldn't be a need for `enableCrossEntityTransactions()`, the reading and forwarding from DLQ can still be done without it, so in that case we don't have to enable this config, you can remove them.

2.	The `maxAutoLockRenewDuration(..)` has no impact on message peeked using "peek*" api. The lock renew matters only if the client is used to receive messages using "receive*" api (those messages which later gets disposition (complete|abandon)). The same goes with `disableAutoComplete()`, but keeping these two doesn't hurt, the peek calls won’t be impacted by these two settings.

--------------------------------------

The message -

> {"az.sdk.message":"AMQP response did not contain OK status code.","statusCode":"NO_CONTENT"} 

means that there is no message in the queue at the moment, in this case, the SDK will return an empty list. So, this is not a problem but service telling the SDK that there is no message to peek.

--------------------------------------

The message "_not opened within.._" means that the SDK cannot establish its TCP connection from the host machine to the Service Bus endpoint. The reason for seeing the multiple occurrences of the error-message

> Error occurred while refreshing token that is not retriable. Not scheduling refresh task

after the connectivity error-message (i.e., "_not opened within.._") is - SDK was trying to establish a TCP connection, but there were also several peekMessages (or other api calls) that were queued up. All of these calls need a shared authentication that is also dependent on the TCP connection. However, SDK could not open the connection after waiting for a minute, so all the peekMessage enqueued that were waiting for auth to complete received signal indicating connectivity issue. This is what the sequence of log messages shows.

In this case, it indicates that there is a temporary network connectivity problem from host running the app to service bus. 

--------------------------------------

Regarding the last question on recommendations – 
1.	One suggestion in terms of coding patterns is - you may use separate builder instance to build the client for each queue (i.e., use 1builder:1queue), rather than the current approach of using a shared builder instance (i.e., not use 1builder:Nqueue). This approach will give more resiliency to your app.
2.	I would also suggest upgrading to 7.15.1. Please refer to this documentation on upgrade and v2 flags to enable. [Troubleshoot Azure Service Bus - Azure SDK for Java | Microsoft Learn](https://learn.microsoft.com/en-us/azure/developer/java/sdk/troubleshooting-messaging-service-bus-overview#upgrade-to-715x)

**Comment by neumannm** [comment]
Hi @anuchandy, thank you for your detailed response and the useful hints! I will adapt my code accordingly.

> The message "not opened within.." means that the SDK cannot establish its TCP connection from the host machine to the Service Bus endpoint. [...] In this case, it indicates that there is a temporary network connectivity problem from host running the app to service bus.

Good to know! Might also explain why I don't see this error when testing locally, only in production (where the application runs in a Kubernetes cluster and the requests go through Azure API management... maybe there's some issue specific to this environment causing connectivity problems 🤔).

Regarding your comment:

> The maxAutoLockRenewDuration(..) has no impact on message peeked using "peek*" api. The lock renew matters only if the client is used to receive messages using "receive*" api (those messages which later gets disposition (complete|abandon)).

-> we also have methods to delete or resend, where we indeed use the receive* api. So what about those - is a `maxAutoLockRenewDuration` of 1 minute advisable (when lock duration is also 1 minute), or should it be higher? I think I read somewhere that these durations should not be the same.

> One question - in the application, are you opening, peeking and closing the client for every http request?

Well yes, for every http request the `peekMessages` Method is called that contains the code snippet from above. Since I'm using try-with-resources, the (`AutoClosable`) receiver client should be closed after the method returns. This is what it looks like after refactoring:

```
try (ServiceBusReceiverClient receiver = new ServiceBusClientBuilder()
			.connectionString(connectionString)
			.configuration(new ConfigurationBuilder()
					.putProperty("com.azure.messaging.servicebus.nonSession.syncReceive.v2", "true")
					.build())
			.receiver()
			.queueName(queueName)
			.buildClient()) {
```

Why are you asking?

**Comment by neumannm** [comment]
Update: I implemented all the suggestions, including updating to 7.15.1 and setting the v2 flag on the receiver client, and also creating a separate builder instance to build the client for each queue (for each request).

Unfortunately, it did not help with the problem. The connection is still lost frequently.

```
WARN 1 --- [ctor-executor-3] c.a.c.a.i.handler.ConnectionHandler: {"az.sdk.message":"onTransportError","connectionId":"MF_5af603_1710239185932","errorCondition":"amqp:connection:framing-error","errorDescription":"org.apache.qpid.proton.engine.TransportException: connection aborted","hostName":"my-sbns.servicebus.windows.net"}
```

I cannot reproduce this when running the application locally (using the same service bus connection), it only happens with the instance running in the Kubernetes cluster. Any ideas how to debug that? I tried doing
`watch nc -z -v my-sbns.servicebus.windows.net 443`
from another container in the same namespace, but this continuously yields
`Connection to my-sbns.servicebus.windows.net (23.102.0.186) 443 port [tcp/https] succeeded!`
even during the time when the timeout occurs.


Sidenote: Each request now gives me a warn message
`WARN 1 --- [nio-8080-exec-1] c.a.m.s.ServiceBusClientBuilder: 'enableAutoComplete' is not supported in synchronous client except through callback receive.`
now that I dropped `disableAutoComplete()`.

**Comment by anuchandy** [comment]
Hi @neumannm, thanks for the additional details. I didn't realize that client creation-disposal happens for every HTTP request. 

It is not an efficient pattern to create and get rid of the client instances for each HTTP request, as this incurs extra costs, such as connection/link negotiation, AD authentication calls and other overhead, impacting the networking. In the common messaging (Service Bus, Event Hubs) scenarios, the client instances last for a long time, so the application should follow a caching approach. The approach here, for example, would be to use _ConcurrentHashMap_ with entry key as the queue name and entry value as client instance. This map will be scoped to the application and is discarded when the application shuts down. An entry could be populated using _computeIfAbsent_ method with a provider that is responsible to new up a builder and client from it. Each time application needs to call peek*, it can reach out to this global map to obtain an already cached client or _computeIfAbsent_ will create one and cache it.

I don't have expertise in AKS infra or it's low-level networking debugging, but I would say, we start with the above approach, which makes the app more suitable for the limited container environments, and then check if it lowers the frequency of the networking connection abort logs that you noticed.

Also, I know that the application has some code for receive/delete/forward as well. If they are in the same pod where peek runs, you can simplify the debugging by temporarily disabling (commenting out) those receive/delete/forward code and only running peek. This will make the logs less cluttered and help you see if the network situation gets better for that peek run. Also disable any other runs in the app or other apps in the same pod.


------------------------

Few questions,
1.	How many queues does the application monitor using the message-peek method you showed earlier?
2.	What is the HTTP request rate looks like – requests / min / sec?

------------------------

Few things to check,
1.	Are the pod (hosting application) and Service Bus namespace deployed to the same region? Client and server located in distinct regions also have an impact on networking. 
2.	I wonder the cores and memory settings for a pod that runs the Java application, too low resourcing (e.g., 0.2, 0.5 milli cores) is another reason for frequent time out or stalling in constrained environment. Based on the case study of many Java containerized production apps, the OpenJDK team at Microsoft suggests no less than 1 core per pod. Here is the documentation by that team reg: core/memory - https://learn.microsoft.com/en-us/azure/developer/java/containers/overview#determine-how-many-cpu-cores-are-needed


------------------------

Also, we can ignore the WARN about auto-complete, it’s a feature available only for ProcessorClient, so SDK is simply saying the chosen client does not support that. You may keep disableAutoComplete (which has no impact on receiver client anyway) if WARN is a noise. But with the caching approach discussed above, we should be seeing this warn only once when a cache entry is populated. 

(Regarding the question on max-renewal, I’ll follow up)

**Comment by anuchandy** [comment]
@neumannm, checking back, did the recommendations above help with your use case?

**Comment by anuchandy** [comment]
Closing this issue, assuming that the suggestions were useful, or this is not a priority at the moment. Feel free to reach out if any assistance is needed at later point.

**Comment by neumannm** [comment]
@anuchandy So sorry for the late reply, project work got me distracted from this issue unfortunately... 

Yes your recommendations seem to have helped. I have changed the code for the peeking receiver as you suggested, using a Map to store receivers per queue, creating each receiver only at first request. Since then, we did not have any notable issues with peeking into our queues.

I still need to refactor the code regarding the functionality to receive from DLQ and re-send to the corresponding "normal" queue for reprocessing. I tried to do the same as with the peeking receiver, but my first attempt was not successful. Which might have to do with the code being suboptimal. You mentioned that cross-entity-transactions are not needed for this functionality, but I was not able to solve it without. I hope I find time soon to dig again into that part.

You said you wanted to follow up on my question regarding max-renewal, did you find anything?

**Comment by anuchandy** [comment]
Hello @neumannm, no worries. Glad to hear that caching pattern worked. 

Regarding the lock-renew for receiver, you’ll need to set an auto-lock renew duration in the client only if the client application is expected to hold on to the received message more than the lock duration set in the entity (queue, topic) level (E.g., in Azure portal).  The client-side lock renewal means calls to the service, so it’s a tradeoff b/w setting a higher value in the portal vs reducing client-side calls to service.

For example, at an entity level if it is currently 60 seconds, but if the application almost always takes ~70 seconds to process and call `complete` then consider bumping entity level value to ~70-80 seconds.
